massa-grpc/src/api.rs

```diff
@@ -0,0 +1,894 @@
+    /// link to the storage component
+    pub storage: Storage,
+    /// gRPC configuration
+    pub grpc_config: GrpcConfig,
+    /// node version
+    pub version: massa_models::version::Version,
+}
+
+impl MassaService {
+    /// Start the gRPC API
+    pub async fn serve(
+        self,
+        config: &GrpcConfig,
+    ) -> Result<StopHandle, Box<dyn std::error::Error>> {
+        let mut svc = GrpcServer::new(self)
+            .max_decoding_message_size(config.max_decoding_message_size)
+            .max_encoding_message_size(config.max_encoding_message_size);
+
+        if let Some(encoding) = &config.accept_compressed {
+            if encoding.eq_ignore_ascii_case("Gzip") {
+                svc = svc.accept_compressed(CompressionEncoding::Gzip);
+            };
+        }
+
+        if let Some(encoding) = &config.accept_compressed {
+            if encoding.eq_ignore_ascii_case("Gzip") {
+                svc = svc.send_compressed(CompressionEncoding::Gzip);
+            };
+        }
+
+        let (shutdown_send, shutdown_recv) = oneshot::channel::<()>();
+
+        if config.accept_http1 {
+            let mut router_with_http1 = tonic::transport::Server::builder()
+                .accept_http1(true)
+                .layer(GrpcWebLayer::new())
+                .add_service(svc);
+
+            if config.enable_reflection {
+                let reflection_service = tonic_reflection::server::Builder::configure()
+                    .register_encoded_file_descriptor_set(FILE_DESCRIPTOR_SET)
+                    .build()?;
+
+                router_with_http1 = router_with_http1.add_service(reflection_service);
+            }
+            router_with_http1
+                .serve_with_shutdown(config.bind, shutdown_recv.map(drop))
+                .await?;
+        } else {
+            let mut router = tonic::transport::Server::builder().add_service(svc.clone());
+
+            if config.enable_reflection {
+                let reflection_service = tonic_reflection::server::Builder::configure()
+                    .register_encoded_file_descriptor_set(FILE_DESCRIPTOR_SET)
+                    .build()?;
+
+                router = router.add_service(reflection_service);
+            }
+
+            router
+                .serve_with_shutdown(config.bind, shutdown_recv.map(drop))
+                .await?;
+        }
+
+        Ok(StopHandle {
+            server_handler: shutdown_send,
+        })
+    }
+}
+
+/// Used to be able to stop the gRPC API
+pub struct StopHandle {
+    server_handler: oneshot::Sender<()>,
+}
+
+impl StopHandle {
+    /// stop the gRPC API gracefully
+    pub fn stop(self) {
+        match self.server_handler.send(()) {
+            Ok(_) => {
+                info!("gRPC API finished cleanly");
+            }
+            Err(err) => warn!("gRPC API thread panicked: {:?}", err),
+        }
+    }
+}
+
+fn match_for_io_error(err_status: &tonic::Status) -> Option<&std::io::Error> {
```

Comment:
todo : we should impl From for specific error and return ours

Reply:
Yeah, the error handling is not done yet